openml/datasets/functions.py

```diff
@@ -463,14 +459,23 @@ def create_dataset(name, description, creator, contributor,
     default_target_attribute : str
         The default target attribute, if it exists.
         Can have multiple values, comma separated.
-    row_id_attribute : str
-        The attribute that represents the row-id column, if present in the dataset.
     ignore_attribute : str | list
         Attributes that should be excluded in modelling, such as identifiers and indexes.
     citation : str
         Reference(s) that should be cited when building on this data.
     version_label : str, optional
         Version label provided by user, can be a date, hash, or some other type of id.
+    row_id_attribute : str, optional
+        The attribute that represents the row-id column, if present in the
+        dataset. If ``data`` is a dataframe and ``row_id_attribute`` is not
+        specified, it will be inferred.
```

Comment:
It would be great if you could be a bit more specific on how it will be inferred. Honestly I'm a bit confused right now. Doesn't a dataframe always have an index?
```
In [1]: import pandas as pd                                                                                                                                                                                  

In [2]: data = [['a'], ['b'], ['c'], ['d'], ['e']]                                                                                                                                                           

In [3]: column_names = ['rnd_str']                                                                                                                                                                           

In [4]: df = pd.DataFrame(data, columns=column_names)                                                                                                                                                        

In [5]: hasattr(data, "index")                                                                                                                                                                               
Out[5]: True
```
Unless the user specifies an index column it should not be uploaded. And I assume that the pandas index column is not uploaded right now.

Reply:
yes, it always has an `index` and the `index.name` is either a string or None. By default, it will be None.

Just to be sure, `row_id_attribute` corresponds to the name of the index column, isn't it?

Reply:
> Just to be sure, row_id_attribute corresponds to the name of the index column, isn't it

Yes, but only if there is one. For most datasets there is none.

> yes, it always has an index and the index.name is either a string or None. By default, it will be None.

Shouldn't the code then be `hasattr(data, "index") and hasattr(data.index, "name")`?

Reply:
> Shouldn't the code then be hasattr(data, "index") and hasattr(data.index, "name")?

No, `index` as always a `name` attribute. It is `None` by default which is the also the value required when `row_id_attribute` should not be specified.

To illustrate what I mean.

```python
In [2]: df = pd.DataFrame([1, 2])

In [3]: df
Out[3]: 
   0
0  1
1  2

In [4]: df.index.name

In [5]: print(df.index.name)
None
```

Reply:
Okay, I got it. Do you think it would make sense to add a brief description about this behavior to the docstring?

Reply:
Python lists also have an index. So it would break for this case.